### PR TITLE
Add support for path parameters in @Route decorator #261

### PR DESCRIPTION
**Pull request by kermitology**

### CODE REVIEW

**Comment by endor** [comment]
Why two commits with the same title?

File: tests/fixtures/inversify/managedController.ts [hunk 1]
```diff
@@ -5,15 +5,15 @@ import { TestModel } from '../testModel';
 import { ManagedService } from './managedService';
 
 @injectable()
-@Route('ManagedTest')
+@Route('{routeParam}/ManagedTest')
```
Comment (endor) [inline tests/fixtures/inversify/managedController.ts]: why in the inversify controller?

File: src/metadataGeneration/parameterGenerator.ts [hunk 2]
```diff
@@ -143,7 +144,24 @@ export class ParameterGenerator {
...
+          default: getInitializerValue(parameter.initializer, type),
+          description: this.getParameterDescription(parameter),
+          in: 'path',
+          name: pathName,
+          parameterName,
+          required: true,
+          type,
+          validators: getParameterValidators(this.parameter, parameterName),
+        };
+      } else {
```
Comment (endor) [inline src/metadataGeneration/parameterGenerator.ts]: why does this need to be inside the routePath check if/else block? can't we handle this before the routePath check?

File: src/metadataGeneration/parameterGenerator.ts [hunk 3]
```diff
@@ -143,7 +144,24 @@ export class ParameterGenerator {
       throw new GenerateMetadataError(`@Path('${parameterName}') Can't support '${type.dataType}' type.`);
     }
     if (!this.path.includes(`{${pathName}}`)) {
-      throw new GenerateMetadataError(`@Path('${parameterName}') Can't match in URL: '${this.path}'.`);
+      if (this.routePath !== undefined ) {
```
Comment (endor) [inline src/metadataGeneration/parameterGenerator.ts]: there's a space too much

File: tests/fixtures/inversify/managedController.ts (hunk 1 above)
Comment (kermitology) [reply tests/fixtures/inversify/managedController.ts]: I just wanted a simple controller to test with. Could be expanded to other tests.

File: src/metadataGeneration/parameterGenerator.ts (hunk 3 above)
Comment (kermitology) [reply src/metadataGeneration/parameterGenerator.ts]: Sorry, my vscode does autoformatting.

File: src/metadataGeneration/methodGenerator.ts [hunk 4]
```diff
@@ -13,6 +13,7 @@ export class MethodGenerator {
 
   constructor(
     private readonly node: ts.MethodDeclaration,
+    private readonly parentPath?: string,
```
[the remark this answers is not in the document]
Comment (kermitology) [reply src/metadataGeneration/methodGenerator.ts]: That would make more sense.

**Comment by endor** [comment]
Now there's no unit test anymore? It would be good to have a unit test. If it's easiest in the inversify controller, then let's put it there. Otherwise this might also justify just creating a new small controller, similar to the `headController`.

**Comment by kermitology** [comment]
Sorry, I haven't done a fork merge like this before, I didn't realize this was committing to the pull request. I'm working on a separate test.

File: tests/fixtures/inversify/routes.ts [hunk 5]
```diff
@@ -110,6 +110,7 @@ export function RegisterRoutes(app: any) {
     authenticateMiddleware([{ "MySecurity": [] }]),
     function(request: any, response: any, next: any) {
       const args={
+        routeParam: { "in": "path", "name": "routeParam", "required": true, "dataType": "string" },
```
Comment (endor) [inline tests/fixtures/inversify/routes.ts]: this shouldn't be in here anymore?

File: tests/fixtures/routeParameter/routeParameterController.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import { inject, injectable } from 'inversify';
```
Comment (endor) [inline tests/fixtures/routeParameter/routeParameterController.ts]: why use inversify in this test controller?

File: tests/integration/route-parameter.spec.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+import * as request from 'supertest';
+import { iocContainer } from '../fixtures/routeParameter/ioc';
+import { RouteParameterService } from '../fixtures/routeParameter/routeParameterService';
+import { app } from '../fixtures/routeParameter/server';
+import { TestModel, TestSubModel } from '../fixtures/testModel';
+
+const basePath = '/v1';
+
+describe('RouteParameter Express Server', () => {
+  it('can handle get request with no path argument', () => {
```
Comment (endor) [inline tests/integration/route-parameter.spec.ts]: why not just add another controller to `fixtures/controllers`, but a whole new "environment" (`fixtures/routeParameter`)?

**Pull request closed without merging**